### PR TITLE
[SMALLFIX] Improve NettyPacketWriter 

### PR DESCRIPTION
**Pull request by apc999**
Various small improvement on `NettyPacketWriter`

- Update callbacks for `sendEOF` and `sendCancel` rather than using `CLOSE_ON_FAILURE`, so that if failures happens on sending the packet to channel, the states (e.g., `mDoneOrFailed` will be updated rather than ignored).
- Better handle exceptions: if there are already exceptions seen by this channel, add them to suppressed.
- Improve Exception message

### CODE REVIEW

File: core/client/fs/src/main/java/alluxio/client/block/stream/NettyPacketWriter.java [hunk 1]
```diff
@@ -348,18 +351,31 @@ public int packetSize() {
...
+   * @param e Exception received
+   */
+  @GuardedBy("mLock")
+  private void updateException(Throwable e) {
+    if (mPacketWriteException == null) {
+      mPacketWriteException = e;
+    } else {
+      mPacketWriteException.addSuppressed(e);
+    }
+  }
```
Comment (calvinjia) [inline core/client/fs/src/main/java/alluxio/client/block/stream/NettyPacketWriter.java]: When will packet write exception be thrown?

File: core/client/fs/src/main/java/alluxio/client/block/stream/NettyPacketWriter.java [hunk 2]
```diff
@@ -188,7 +188,8 @@ public void writePacket(final ByteBuf buf) throws IOException {
         try {
           if (!mBufferNotFullOrFailed.await(WRITE_TIMEOUT_MS, TimeUnit.MILLISECONDS)) {
             throw new DeadlineExceededException(
-                String.format("Timeout writing to %s for request %s.", mAddress, mPartialRequest));
+                String.format("Timeout writing to %s for request %s after %d ms.",
```
[the remark this answers is not in the document]
Comment (apc999) [reply core/client/fs/src/main/java/alluxio/client/block/stream/NettyPacketWriter.java]: fixed

File: core/client/fs/src/main/java/alluxio/client/block/stream/NettyPacketWriter.java [hunk 3]
```diff
@@ -262,17 +264,18 @@ public void run() {
             });
             throw new UnavailableException(mPacketWriteException);
           }
-          if (!mDoneOrFailed
-              .await(Configuration.getMs(PropertyKey.USER_NETWORK_NETTY_WRITER_CLOSE_TIMEOUT_MS),
-                  TimeUnit.MILLISECONDS)) {
+          long timeoutMs =
```
[the remark this answers is not in the document]
Comment (apc999) [reply core/client/fs/src/main/java/alluxio/client/block/stream/NettyPacketWriter.java]: good call. fixed

File: core/client/fs/src/main/java/alluxio/client/block/stream/NettyPacketWriter.java [hunk 4]
```diff
@@ -461,5 +477,30 @@ public void operationComplete(ChannelFuture future) {
...
+      if (!future.isSuccess()) {
+        future.channel().close();
+      }
+      try (LockResource lr = new LockResource(mLock)) {
+        if (future.cause() != null) {
+          updateException(future.cause());
+          mDoneOrFailed.signal();
+          mBufferNotFullOrFailed.signal();
+          mBufferEmptyOrFailed.signal();
+        }
```
[the remark this answers is not in the document]
Comment (apc999) [reply core/client/fs/src/main/java/alluxio/client/block/stream/NettyPacketWriter.java]: from the javadoc of `ChannelFuture`,  `future.cause() != null` is ensured if there is a failure.

[the remark this answers is not in the document]
Comment (calvinjia) [reply core/client/fs/src/main/java/alluxio/client/block/stream/NettyPacketWriter.java]: Can we use the same conditions?

File: core/client/fs/src/main/java/alluxio/client/block/stream/NettyPacketWriter.java [hunk 5]
```diff
@@ -262,17 +266,16 @@ public void run() {
...
+          if (!mDoneOrFailed.await(CLOSE_TIMEOUT_MS, TimeUnit.MILLISECONDS)) {
             closeFuture = mChannel.eventLoop().submit(new Runnable() {
               @Override
               public void run() {
                 mChannel.close();
               }
             });
             throw new DeadlineExceededException(String.format(
-                "Timeout closing PacketWriter to %s for request %s.", mAddress, mPartialRequest));
+                "Timeout closing PacketWriter to %s for request %s after %s ms.",
```
Comment (calvinjia) [inline core/client/fs/src/main/java/alluxio/client/block/stream/NettyPacketWriter.java]: `%dms`?

**Comment by calvinjia** [comment]
@apc999 had 2 follow up questions.

File: core/client/fs/src/main/java/alluxio/client/block/stream/NettyPacketWriter.java (hunk 5 above)
Comment (apc999) [reply core/client/fs/src/main/java/alluxio/client/block/stream/NettyPacketWriter.java]: fixed

File: core/client/fs/src/main/java/alluxio/client/block/stream/NettyPacketWriter.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (apc999) [reply core/client/fs/src/main/java/alluxio/client/block/stream/NettyPacketWriter.java]: fixed

**Comment by apc999** [comment]
@calvinjia PTAL

**Pull request merged**